### ISSUE TITLE
Show deprecated template items in IDE

### ISSUE THREAD

**Issue by eliasm307** [angular/angular#59343]
### Which @angular/* package(s) are relevant/related to the feature request?

language-service

### Description

This issue was raised in the [vscode-ng-language-service](https://github.com/angular/vscode-ng-language-service) repo by  [@mr1008](https://github.com/Mr1008) ie https://github.com/angular/vscode-ng-language-service/issues/1995, however I think it would be implemented in the base language service which is in this repo.

Description copied from https://github.com/angular/vscode-ng-language-service/issues/1995,

> # 🚀 feature request
> ### Description
> Currently there is no way of marking components as deprecated. It would be cool if template editor supported standard way how deprecated code should be marked in TS/JS, using JSDoc [@deprecated](https://github.com/deprecated) feature. This would help working in enterprise grade environment when there are multiple components and some of these should no longer be used. Currently it is very easy to forget/overlook such usage. Displaying it somehow in editor in straightforward way would help a lot.
> 
> ### Feature Type
> Display somehow that component is deprecated.
> 
> ### Describe the solution you'd like
> I believe the standard way of displaying would be fine, similar way as vscode does for TS/JS code. Currently WebStorm supports it and it looks like in the screenshot: ![image](https://github.com/angular/vscode-ng-language-service/assets/68107706/96a3d6f6-725c-4cf6-84b6-dd5074d50531)
> 
> It supports standard way of marking elements as deprecated using JSDoc [@deprecated](https://github.com/deprecated) feature. For example:
> 
> /**
>  * @deprecated DO NOT USE THIS TABLE.
>  */
> @Component({
>   selector: 'app-table-old',
>   templateUrl: './old-table.component.html',
>   styleUrls: ['./old-table.component.scss']
> })

FYI @Mr1008

### Proposed solution

This could be implemented as a rule that provides deprecated diagnostics, similar to as was done for the `unused imports` rule ie https://github.com/angular/angular/issues/46766

### Alternatives considered

No alternatives in VS Code.

The functionality seems to be available in Webstorm, sounds like they have a forked or extended version of the language service. It would be good to have this in the base language service so more IDEs can have this functionality.

**Comment by eneajaho** [comment]
Just to note it down, when using standalone components it will show up in the imports array as deprecated.

![Image](https://github.com/user-attachments/assets/8220d8e6-9d9b-4a37-95c6-cf496f1cc69b)

**Comment by eliasm307** [comment]
@eneajaho

Yes that's in typescript files, as mentioned there no issue there, but if just looking at the component template (eg in a different file) it's not clear what is deprecated, so devs could possibly keep using deprecated things/APIs without realising

**Comment by eneajaho** [comment]
https://github.com/angular/angular/pull/59524

Looks like a fix is coming.

**Comment by atscott** [comment]
> https://github.com/angular/angular/pull/59524

> Looks like a fix is coming

Not quite.  This would need to be done by providing semantic tokens, following this guide https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide.

One quick idea would be to get the semantic tokens for the type check block from the TS language service, filter down to the deprecated ones, and then map those back to their template locations.

**Comment by eliasm307** [comment]
@atscott from the investigation I did, I think also providing diagnostics would work, as you can add a diagnostic tag which produces the strike through effect, atleast on vs code.

Benefit of doing it as a diagnostic is that they are created and pushed from the base language service, which means this would apply to any lsp editors that use the language service

Also, regarding the comment about type check blocks, sounds like that would be specific to the js like aspects of templates, which would be useful, but wouldn't cover the entire template eg pipes, directives, components/elements etc which was the main motivation for the issue

**Comment by ivanwonder** [comment]
> [@atscott](https://github.com/atscott) from the investigation I did, I think also providing diagnostics would work, as you can add a diagnostic tag which produces the strike through effect, atleast on vs code.

I agree to provide diagnostics for the deprecated information. I try using TCB. The challenging part is that the component class is used as a type, but there is no position map information for it.

<img width="1192" alt="Image" src="https://github.com/user-attachments/assets/cb35a01f-cb40-4ad8-b18a-e9e52109c9fa" />

**Comment by atscott** [comment]
> but there is no position map information for it.

The place screenshot points to wouldn’t have a mapping to a template location because the constructor definition could be used in several places in the template. Its usage does have the source map comment below (var _t2).

**Comment by ivanwonder** [comment]
> > but there is no position map information for it.
> 
> The place screenshot points to wouldn’t have a mapping to a template location because the constructor definition could be used in several places in the template. Its usage does have the source map comment below (var _t2).

Yes, but the deprecated range the ts reports is for the constructor definition, need to find a way to link it to the _t2.
